{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "67c84f11-c21f-432c-bafe-be2febccf36f",
   "metadata": {},
   "source": [
    "TODO:\n",
    "\n",
    "- Experiment with using a simpler empty pipeline instead of en_core_web_sm or disable unwanted pipelines, see *Disabling pipeline components* in chapter 3 of learn-spaCy project\n",
    "- Use *components with extensions* from chapter 3 of learn-spaCy project to look up the model category information e.g. training type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "de4028b5-0817-451a-a4a5-97139e5fd3de",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import spacy\n",
    "import PyPDF2\n",
    "import fitz  # PyMuPDF\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "cff926e1-2f6e-402f-ae41-35bdb400fd7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "ROOT_FOLDER = '../../../'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "51370296-6d37-4310-a92d-11d0999fb605",
   "metadata": {},
   "outputs": [],
   "source": [
    "DATA_FOLDER = '../../../data/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "b4b899a7-16b4-4e27-8d1d-5049c3de355c",
   "metadata": {},
   "outputs": [],
   "source": [
    "DOWNLOAD_FOLDER = '../../../downloads/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "a4f6ab4c-80cd-4230-892a-0994f2b3139d",
   "metadata": {},
   "outputs": [],
   "source": [
    "large_ml_lncRNA_search_df = pd.read_parquet(f'{DATA_FOLDER}large_ml_lncRNA_search_df.parquet')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dba1c80a-52b2-48db-a708-ad05ea26a784",
   "metadata": {},
   "source": [
    "Filter for files that are in English"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "1fbdc192-03b4-4973-a100-c87cbd069728",
   "metadata": {},
   "outputs": [],
   "source": [
    "large_ml_lncRNA_search_df = large_ml_lncRNA_search_df[large_ml_lncRNA_search_df['language'] == 'en']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d4fa79b0-de16-415c-aa30-c89ff7827288",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>title</th>\n",
       "      <th>abstract</th>\n",
       "      <th>year</th>\n",
       "      <th>url</th>\n",
       "      <th>author_id</th>\n",
       "      <th>query</th>\n",
       "      <th>file_name</th>\n",
       "      <th>file_path</th>\n",
       "      <th>language</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>LncMachine: a machine learning algorithm for l...</td>\n",
       "      <td>We evaluated the performance of machine learni...</td>\n",
       "      <td>2021</td>\n",
       "      <td>https://escholarship.org/content/qt32n7m7td/qt...</td>\n",
       "      <td>[ZeGca3cAAAAJ, o3DdNZMAAAAJ, Ydo9ResAAAAJ, 4IR...</td>\n",
       "      <td>Machine Learning lncRNA</td>\n",
       "      <td>1fcc8c69-cc2d-48d1-9052-81b0154706cd.pdf</td>\n",
       "      <td>../../../downloads/1fcc8c69-cc2d-48d1-9052-81b...</td>\n",
       "      <td>en</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>DMFLDA: a deep learning framework for predicti...</td>\n",
       "      <td>lncRNAs and diseases, and traditional machine ...</td>\n",
       "      <td>2020</td>\n",
       "      <td>https://minzeng1990.github.io/Files/TCBB-DMFLD...</td>\n",
       "      <td>[Q6b80i8AAAAJ, wicacBwAAAAJ, kERS9vUAAAAJ, O_x...</td>\n",
       "      <td>Machine Learning lncRNA</td>\n",
       "      <td>e1460d83-589f-45ab-93fc-8b36160efa8d.pdf</td>\n",
       "      <td>../../../downloads/e1460d83-589f-45ab-93fc-8b3...</td>\n",
       "      <td>en</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>Evaluation of deep learning in non-coding RNA ...</td>\n",
       "      <td>In this study, we review the progress of ncRNA...</td>\n",
       "      <td>2019</td>\n",
       "      <td>https://u.osu.edu/bmbl/files/2021/01/lncfinder...</td>\n",
       "      <td>[xAqx-WkAAAAJ, 2tbe1RoAAAAJ, Vt5edEkAAAAJ]</td>\n",
       "      <td>Machine Learning lncRNA</td>\n",
       "      <td>59adb036-a0ef-4839-ab89-4f3990b7e26e.pdf</td>\n",
       "      <td>../../../downloads/59adb036-a0ef-4839-ab89-4f3...</td>\n",
       "      <td>en</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>A review of machine learning-based prediction ...</td>\n",
       "      <td>the subcellular localization of lncRNAs on a l...</td>\n",
       "      <td>2023</td>\n",
       "      <td>http://www.clausiuspress.com/assets/default/ar...</td>\n",
       "      <td>[, ]</td>\n",
       "      <td>Machine Learning lncRNA</td>\n",
       "      <td>b0d97201-c1f7-435d-9f82-0a36c76a004f.pdf</td>\n",
       "      <td>../../../downloads/b0d97201-c1f7-435d-9f82-0a3...</td>\n",
       "      <td>en</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>LncRNA Subcellular Localization Signals–Are th...</td>\n",
       "      <td>at the 5’ end or 3’ end of lncRNA [2]. In this...</td>\n",
       "      <td>2023</td>\n",
       "      <td>https://par.nsf.gov/servlets/purl/10538687</td>\n",
       "      <td>[Lcmc_iUAAAAJ, L8nlPxYAAAAJ, fEQEjCIAAAAJ, Wfz...</td>\n",
       "      <td>Machine Learning lncRNA</td>\n",
       "      <td>7889091f-b484-4e8d-bab2-0b150ba85ce3.pdf</td>\n",
       "      <td>../../../downloads/7889091f-b484-4e8d-bab2-0b1...</td>\n",
       "      <td>en</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>913</th>\n",
       "      <td>Mitochondrial Import of Malat1 Regulates Cardi...</td>\n",
       "      <td>machine learning in the classi cation of lncRN...</td>\n",
       "      <td>2020</td>\n",
       "      <td>https://scholar.archive.org/work/urt26oavknafp...</td>\n",
       "      <td>[hhOh95IAAAAJ, SWozBjAAAAAJ, , X_RVrq4AAAAJ]</td>\n",
       "      <td>Machine Learning lncRNA</td>\n",
       "      <td>a8d9deec-c217-4bf1-86a8-c4bb8c10a6a2.pdf</td>\n",
       "      <td>../../../downloads/a8d9deec-c217-4bf1-86a8-c4b...</td>\n",
       "      <td>en</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>917</th>\n",
       "      <td>Impact of sequencing technologies on long non-...</td>\n",
       "      <td>In summary, the analyzed tools use supervised ...</td>\n",
       "      <td>2022</td>\n",
       "      <td>https://www.biorxiv.org/content/biorxiv/early/...</td>\n",
       "      <td>[, , ]</td>\n",
       "      <td>Machine Learning lncRNA</td>\n",
       "      <td>3ce36b96-ff84-4b84-9c17-f6fbdbbf1cef.pdf</td>\n",
       "      <td>../../../downloads/3ce36b96-ff84-4b84-9c17-f6f...</td>\n",
       "      <td>en</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>921</th>\n",
       "      <td>Deciphering the methylation landscape in breas...</td>\n",
       "      <td>Innovative automated machine learning was empl...</td>\n",
       "      <td>2021</td>\n",
       "      <td>http://repository-empedu-rd.ekt.gr/empedu-rd/b...</td>\n",
       "      <td>[, hojAa00AAAAJ]</td>\n",
       "      <td>Machine Learning lncRNA</td>\n",
       "      <td>48c52af8-3a4f-40b9-91a8-dd04c8a41f1d.pdf</td>\n",
       "      <td>../../../downloads/48c52af8-3a4f-40b9-91a8-dd0...</td>\n",
       "      <td>en</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>928</th>\n",
       "      <td>Machine learning models for predicting lymph n...</td>\n",
       "      <td>These analyses should be extended to the effec...</td>\n",
       "      <td>2020</td>\n",
       "      <td>https://scholar.archive.org/work/bbbmcpqig5eyp...</td>\n",
       "      <td>[, , , , , ]</td>\n",
       "      <td>Machine Learning lncRNA</td>\n",
       "      <td>b55db42f-8ffb-452b-81be-a8e4270ef0a3.pdf</td>\n",
       "      <td>../../../downloads/b55db42f-8ffb-452b-81be-a8e...</td>\n",
       "      <td>en</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>929</th>\n",
       "      <td>Development of New Bioinformatic Approaches fo...</td>\n",
       "      <td>(lncRNAs) in ID due to their role in gene expr...</td>\n",
       "      <td>2017</td>\n",
       "      <td>https://tigerprints.clemson.edu/cgi/viewconten...</td>\n",
       "      <td>[ge53c0kAAAAJ]</td>\n",
       "      <td>Machine Learning lncRNA</td>\n",
       "      <td>7ea894e6-f571-456b-a4db-2c9997acf44b.pdf</td>\n",
       "      <td>../../../downloads/7ea894e6-f571-456b-a4db-2c9...</td>\n",
       "      <td>en</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>76 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 title  \\\n",
       "12   LncMachine: a machine learning algorithm for l...   \n",
       "29   DMFLDA: a deep learning framework for predicti...   \n",
       "48   Evaluation of deep learning in non-coding RNA ...   \n",
       "58   A review of machine learning-based prediction ...   \n",
       "89   LncRNA Subcellular Localization Signals–Are th...   \n",
       "..                                                 ...   \n",
       "913  Mitochondrial Import of Malat1 Regulates Cardi...   \n",
       "917  Impact of sequencing technologies on long non-...   \n",
       "921  Deciphering the methylation landscape in breas...   \n",
       "928  Machine learning models for predicting lymph n...   \n",
       "929  Development of New Bioinformatic Approaches fo...   \n",
       "\n",
       "                                              abstract  year  \\\n",
       "12   We evaluated the performance of machine learni...  2021   \n",
       "29   lncRNAs and diseases, and traditional machine ...  2020   \n",
       "48   In this study, we review the progress of ncRNA...  2019   \n",
       "58   the subcellular localization of lncRNAs on a l...  2023   \n",
       "89   at the 5’ end or 3’ end of lncRNA [2]. In this...  2023   \n",
       "..                                                 ...   ...   \n",
       "913  machine learning in the classi cation of lncRN...  2020   \n",
       "917  In summary, the analyzed tools use supervised ...  2022   \n",
       "921  Innovative automated machine learning was empl...  2021   \n",
       "928  These analyses should be extended to the effec...  2020   \n",
       "929  (lncRNAs) in ID due to their role in gene expr...  2017   \n",
       "\n",
       "                                                   url  \\\n",
       "12   https://escholarship.org/content/qt32n7m7td/qt...   \n",
       "29   https://minzeng1990.github.io/Files/TCBB-DMFLD...   \n",
       "48   https://u.osu.edu/bmbl/files/2021/01/lncfinder...   \n",
       "58   http://www.clausiuspress.com/assets/default/ar...   \n",
       "89          https://par.nsf.gov/servlets/purl/10538687   \n",
       "..                                                 ...   \n",
       "913  https://scholar.archive.org/work/urt26oavknafp...   \n",
       "917  https://www.biorxiv.org/content/biorxiv/early/...   \n",
       "921  http://repository-empedu-rd.ekt.gr/empedu-rd/b...   \n",
       "928  https://scholar.archive.org/work/bbbmcpqig5eyp...   \n",
       "929  https://tigerprints.clemson.edu/cgi/viewconten...   \n",
       "\n",
       "                                             author_id  \\\n",
       "12   [ZeGca3cAAAAJ, o3DdNZMAAAAJ, Ydo9ResAAAAJ, 4IR...   \n",
       "29   [Q6b80i8AAAAJ, wicacBwAAAAJ, kERS9vUAAAAJ, O_x...   \n",
       "48          [xAqx-WkAAAAJ, 2tbe1RoAAAAJ, Vt5edEkAAAAJ]   \n",
       "58                                                [, ]   \n",
       "89   [Lcmc_iUAAAAJ, L8nlPxYAAAAJ, fEQEjCIAAAAJ, Wfz...   \n",
       "..                                                 ...   \n",
       "913       [hhOh95IAAAAJ, SWozBjAAAAAJ, , X_RVrq4AAAAJ]   \n",
       "917                                             [, , ]   \n",
       "921                                   [, hojAa00AAAAJ]   \n",
       "928                                       [, , , , , ]   \n",
       "929                                     [ge53c0kAAAAJ]   \n",
       "\n",
       "                       query                                 file_name  \\\n",
       "12   Machine Learning lncRNA  1fcc8c69-cc2d-48d1-9052-81b0154706cd.pdf   \n",
       "29   Machine Learning lncRNA  e1460d83-589f-45ab-93fc-8b36160efa8d.pdf   \n",
       "48   Machine Learning lncRNA  59adb036-a0ef-4839-ab89-4f3990b7e26e.pdf   \n",
       "58   Machine Learning lncRNA  b0d97201-c1f7-435d-9f82-0a36c76a004f.pdf   \n",
       "89   Machine Learning lncRNA  7889091f-b484-4e8d-bab2-0b150ba85ce3.pdf   \n",
       "..                       ...                                       ...   \n",
       "913  Machine Learning lncRNA  a8d9deec-c217-4bf1-86a8-c4bb8c10a6a2.pdf   \n",
       "917  Machine Learning lncRNA  3ce36b96-ff84-4b84-9c17-f6fbdbbf1cef.pdf   \n",
       "921  Machine Learning lncRNA  48c52af8-3a4f-40b9-91a8-dd04c8a41f1d.pdf   \n",
       "928  Machine Learning lncRNA  b55db42f-8ffb-452b-81be-a8e4270ef0a3.pdf   \n",
       "929  Machine Learning lncRNA  7ea894e6-f571-456b-a4db-2c9997acf44b.pdf   \n",
       "\n",
       "                                             file_path language  \n",
       "12   ../../../downloads/1fcc8c69-cc2d-48d1-9052-81b...       en  \n",
       "29   ../../../downloads/e1460d83-589f-45ab-93fc-8b3...       en  \n",
       "48   ../../../downloads/59adb036-a0ef-4839-ab89-4f3...       en  \n",
       "58   ../../../downloads/b0d97201-c1f7-435d-9f82-0a3...       en  \n",
       "89   ../../../downloads/7889091f-b484-4e8d-bab2-0b1...       en  \n",
       "..                                                 ...      ...  \n",
       "913  ../../../downloads/a8d9deec-c217-4bf1-86a8-c4b...       en  \n",
       "917  ../../../downloads/3ce36b96-ff84-4b84-9c17-f6f...       en  \n",
       "921  ../../../downloads/48c52af8-3a4f-40b9-91a8-dd0...       en  \n",
       "928  ../../../downloads/b55db42f-8ffb-452b-81be-a8e...       en  \n",
       "929  ../../../downloads/7ea894e6-f571-456b-a4db-2c9...       en  \n",
       "\n",
       "[76 rows x 9 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "large_ml_lncRNA_search_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "47e96863-45ff-4fb3-abc9-b2af9bec3ae1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "fa1d524a-861b-4408-9691-39142777bc31",
   "metadata": {},
   "outputs": [],
   "source": [
    "paper_titles = large_ml_lncRNA_search_df['title'].values.tolist()\n",
    "paper_pdf_file_paths = large_ml_lncRNA_search_df['file_path'].values.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e80ce6b1-d020-4888-8e8b-48b76b79839c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "c2d1199d-4667-4389-8b39-192c24d133ae",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<spacy.pipeline.entityruler.EntityRuler at 0x7d139e064650>"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import spacy\n",
    "from spacy.matcher import PhraseMatcher\n",
    "\n",
    "# Load the EntityRuler patterns from a file\n",
    "nlp = spacy.load(\"en_core_web_sm\")\n",
    "ruler = nlp.add_pipe(\"entity_ruler\", before=\"ner\")\n",
    "ruler.from_disk(f\"{DATA_FOLDER}ml_entity_ruler_patterns\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "184a9525-b193-4a12-9630-5a63a9e7c511",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Support Vector Machine ML_METHOD SUPPORT_VECTOR_MACHINES\n"
     ]
    }
   ],
   "source": [
    "# Test the pipeline\n",
    "text = \"Support Vector Machine, Support Vector Machines, SVM, and S.V.M. are popular machine learning methods.\"\n",
    "doc = nlp(text)\n",
    "\n",
    "# Print detected entities\n",
    "for ent in doc.ents:\n",
    "    if ent.label_ == \"ML_METHOD\":\n",
    "        print(ent.text, ent.label_, ent.id_)\n",
    "        break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "f9efe7d3-6abd-40bd-b6b8-ace8771bc3d5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Support Vector Machine'"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ent.text"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "9f3e0b05-cd7d-43d8-bed5-981006abd2c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_sections_from_pdf(pdf_file_path):\n",
    "    \"\"\"\n",
    "    Extracts sections from a PDF and identifies specific sections like Introduction and Discussion,\n",
    "    while ignoring References and Appendices.\n",
    "\n",
    "    Args:\n",
    "        pdf_file_path (str): Path to the PDF file.\n",
    "\n",
    "    Returns:\n",
    "        dict: A dictionary with section titles as keys and their content as values.\n",
    "    \"\"\"\n",
    "    # Open the PDF file\n",
    "    doc = fitz.open(pdf_file_path)\n",
    "    \n",
    "    # Initialize variables\n",
    "    sections = {}\n",
    "    current_section = None\n",
    "    current_text = []\n",
    "\n",
    "    # Regular expressions to identify sections\n",
    "    section_pattern = re.compile(r\"^(Introduction|Discussion|Conclusion|Methods|Results)$\", re.IGNORECASE)\n",
    "    #section_pattern = re.compile(r\"^(Introduction|Discussion|Conclusion|Methods|Results|Appendix|Appendices)$\", re.IGNORECASE)\n",
    "    ignore_pattern = re.compile(r\"^(References|Appendix|Appendices)$\", re.IGNORECASE)\n",
    "    #ignore_pattern = re.compile(r\"^(References)$\", re.IGNORECASE)\n",
    "\n",
    "    for page in doc:\n",
    "        # Extract text from the page\n",
    "        text = page.get_text(\"text\")\n",
    "\n",
    "        # Split text into lines for processing\n",
    "        lines = text.splitlines()\n",
    "\n",
    "        for line in lines:\n",
    "            line = line.strip()\n",
    "            if not line:\n",
    "                continue\n",
    "\n",
    "            # Check if the line matches a section title\n",
    "            section_match = section_pattern.match(line)\n",
    "            ignore_match = ignore_pattern.match(line)\n",
    "\n",
    "            if section_match:\n",
    "                # Save the current section's text\n",
    "                if current_section and current_text:\n",
    "                    sections[current_section] = \"\\n\".join(current_text)\n",
    "\n",
    "                # Start a new section\n",
    "                current_section = section_match.group(0).capitalize()\n",
    "                current_text = []\n",
    "\n",
    "            elif ignore_match:\n",
    "                # Stop processing further sections if references or appendices are encountered\n",
    "                if current_section and current_text:\n",
    "                    sections[current_section] = \"\\n\".join(current_text)\n",
    "                current_section = None\n",
    "                current_text = []\n",
    "\n",
    "            elif current_section:\n",
    "                # Append text to the current section\n",
    "                current_text.append(line)\n",
    "\n",
    "    # Save the last section's text\n",
    "    if current_section and current_text:\n",
    "        sections[current_section] = \"\\n\".join(current_text)\n",
    "\n",
    "    doc.close()\n",
    "    return sections"
   ]
  },
  {
   "cell_type": "raw",
   "id": "58487727-1fcd-4dc1-bb5c-419ff3c49ec1",
   "metadata": {},
   "source": [
    "def search_in_file(pdf_file_path):    \n",
    "    search_results = dict()\n",
    "    # Extract text from PDF\n",
    "    with open(pdf_file_path, \"rb\") as pdf_file:\n",
    "        pdf_reader = PyPDF2.PdfReader(pdf_file)\n",
    "        for page_num, page in enumerate(pdf_reader.pages):\n",
    "            # Extract text from the page\n",
    "            text = page.extract_text()\n",
    "            if text:\n",
    "                # Use spaCy to process the text\n",
    "                doc = nlp(text)\n",
    "                \n",
    "                # Detect entities\n",
    "                for ent in doc.ents:\n",
    "                    if ent.label_ == \"ML_METHOD\":\n",
    "                        if ent.id_ in search_results:\n",
    "                            search_results[ent.id_].append(page_num + 1)  # Store page number (1-indexed)\n",
    "                        else:\n",
    "                            search_results[ent.id_] = [page_num + 1]  # Store page number (1-indexed)\n",
    "    \n",
    "    # Display search results\n",
    "    for term, pages in search_results.items():\n",
    "        if pages:\n",
    "            print(f\"'{term}' found on page(s): {set(pages)}\")\n",
    "        #else:\n",
    "        #    print(f\"'{term}' not found in the document.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "a1a7d9c0-9d94-4bdb-9c57-6a53f97062f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def search_in_file(pdf_file_path):    \n",
    "    search_results = dict()\n",
    "    # Extract text from PDF\n",
    "    with open(pdf_file_path, \"rb\") as pdf_file:\n",
    "        pdf_reader = PyPDF2.PdfReader(pdf_file)\n",
    "        for page_num, page in enumerate(pdf_reader.pages):\n",
    "            # Extract text from the page\n",
    "            text = page.extract_text()\n",
    "            if text:\n",
    "                # Use spaCy to process the text\n",
    "                doc = nlp(text)\n",
    "                \n",
    "                # Detect entities\n",
    "                for ent in doc.ents:\n",
    "                    if ent.label_ == \"ML_METHOD\":\n",
    "                        if ent.id_ in search_results:\n",
    "                            search_results[ent.id_] += 1  # Increment Count\n",
    "                        else:\n",
    "                            search_results[ent.id_] = 1 # Store count\n",
    "    \n",
    "    return search_results"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "92ef748d-a583-4a24-a60b-6600fb596901",
   "metadata": {},
   "source": [
    "Search through all papers and extract ML_METHOD info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "c8b9b31f-1f40-4834-8817-a3168617cf11",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "processing: LncMachine: a machine learning algorithm for long noncoding RNA annotation in plants\n",
      "processing: DMFLDA: a deep learning framework for predicting lncRNA–disease associations\n",
      "processing: Evaluation of deep learning in non-coding RNA classification\n",
      "processing: A review of machine learning-based prediction of lncRNA subcellular localization\n",
      "processing: LncRNA Subcellular Localization Signals–Are the Two Ends Equal? A Machine Learning Analysis Across Multiple Cell Lines\n",
      "processing: Identification of Characteristic lncRNA Molecular Markers in Osteoarthritis by Integrating GEO Database and Machine Learning Strategies and Experimental Validation\n",
      "processing: Towards machine learning in molecular biology\n",
      "processing: SDLDA: lncRNA-disease association prediction based on singular value decomposition and deep learning\n",
      "processing: A systematic evaluation of the computational tools for lncRNA identification\n",
      "processing: LncDLSM: identification of long non-coding RNAs with deep learning-based sequence model\n",
      "processing: ncRNAInter: a novel strategy based on graph neural network to discover interactions between lncRNA and miRNA\n",
      "processing: Exploring the Genetic and Transcriptomic Landscape of Obsessive-Compulsive Disorder: A Machine Learning-Driven Investigation of lncRNA BDNF-AS\n",
      "processing: Computational Prediction of lncRNA-Protein Interactions using Machine learning\n",
      "processing: Using machine learning to predict long non-coding RNAs and exploring their evolutionary patterns and prevalence in plant transcriptomes\n",
      "processing: DeepLncLoc: a deep learning framework for long non-coding RNA subcellular localization prediction based on subsequence embedding\n",
      "processing: DBNLDA: Deep Belief Network based representation learning for lncRNA-disease association prediction\n",
      "processing: Anoikis-related mRNA-lncRNA and DNA methylation profiles for overall survival prediction in breast cancer patients\n",
      "processing: DeepHEN: quantitative prediction essential lncRNA genes and rethinking essentialities of lncRNA genes\n",
      "processing: RLF-LPI: An ensemble learning framework using sequence information for predicting lncRNA-protein interaction based on AE-ResLSTM and fuzzy decision\n",
      "processing: Machine learning-based state-of-the-art methods for the classification of rna-seq data\n",
      "processing: Exploring neural network models for LncRNA sequence identification\n",
      "processing: A deep learning approach to lncRNA subcellular localization using inexact q-mers\n",
      "processing: FLYNC: A Machine Learning-Driven Framework for Discovering Long Non-Coding RNAs in Drosophila melanogaster\n",
      "processing: Secondary structure prediction of long noncoding RNA: review and experimental comparison of existing approaches\n",
      "processing: Long non-coding RNA functional annotation: machine learning approaches\n",
      "processing: A review of computational model-based prediction of lncRNA subcellular localization\n",
      "processing: Multi-omics characterization and machine learning of lung adenocarcinoma molecular subtypes to guide precise chemotherapy and immunotherapy\n",
      "processing: MLCDForest: multi-label classification with deep forest in disease prediction for long non-coding RNAs\n",
      "processing: Deepcnpp: deep learning architecture to distinguish the promoter of human long non-coding rna genes and protein-coding genes\n",
      "processing: Feature Selection and Classification Reveal Key lncRNAs for Multiple Cancers\n",
      "processing: Classification of non-coding rna-a review from machine learning perspective\n",
      "processing: … synthesis-associated pivotal biomarkers identification and ceRNA networks construction in frozen shoulder by comprehensive bioinformatics and machine learning …\n",
      "processing: Heterogeneous network and graph attention auto-encoder for LncRNA-disease association prediction\n",
      "processing: LncRNA–disease association prediction through combining linear and non-linear features with matrix factorization and deep learning techniques\n",
      "processing: Bioinformatics tools for lncRNA research\n",
      "processing: LncRNA-disease association prediction method based on heterogeneous information completion and convolutional neural network\n",
      "processing: Identification of a lncRNA-mRNA Regulatory Module for Prognosis, Classification and Potential Treatment of Lung Adenocarcinoma\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "incorrect startxref pointer(2)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "processing: An Optimized Technique for RNA Prediction Based on Neural Network\n",
      "processing: A workflow combining single-cell CRISPRi screening and a supervised autoencoder neural network to detect subtle transcriptomic perturbations induced by lncRNA …\n",
      "processing: LncRNA FAM83H-AS1 Amplification is Associated With a Poor Prognosis in Lung Adenocarcinoma and Can Serve as A Therapeutic Target\n",
      "processing: Machine learning-informed liquid-liquid phase separation for personalized breast cancer treatment assessment\n",
      "processing: HRGCNLDA: Forecasting of lncRNA-disease association based on hierarchical refinement graph convolutional neural network\n",
      "processing: Comparison of performance of different k values with k-fold cross validation in a graph-based learning model for incrna-disease prediction\n",
      "processing: GATLncLoc+ C&S: Prediction of LncRNA subcellular localization based on corrective graph attention network\n",
      "processing: Identification of Candidate Diagnostic gene CDC25C for Trametinib-induced Cardiotoxicity by Integrating Bioinformatics and Machine Learning\n",
      "processing: LncPNdeep: A long non-coding RNA classifier based on Large Language Model with peptide and nucleotide embedding\n",
      "processing: A Review on Genomics Data Analysis using Machine Learning\n",
      "processing: MixLMI: A Multi-level Enhancement Model for Plant lncRNA-miRNA Interaction Prediction\n",
      "processing: Flawed machine-learning confounds coding sequence annotation\n",
      "processing: Optimized Class Level Attraction Measure Based Machine Learning for Improved Disease Prediction Using Support Vector Machine\n",
      "processing: Machine learning revealed novel ferroptosis-related genes and construction ceRNA network in dermal lymphatic endothelial cells of diabetic foot ulcer\n",
      "processing: A Machine Learning Framework for Identifying Molecular Biomarkers from Transcriptomic Cancer Data\n",
      "processing: Unraveling lncRNA Diversity at a Single Cell Resolution and in a Spatial Context across Different Cancer Types\n",
      "processing: Predict lncRNA-disease associations based on graphlet interaction\n",
      "processing: Identification of the Molecular Subtypes of Acute Ischemic Stroke Using Bioinformatics and Machine Learning\n",
      "processing: Discovering coding LNCRNAS using deep learning training dynamics\n",
      "processing: Investigating The Roles of microRNAs/lncRNAs in Characterizing Breast Cancer Subtypes and Prognosis\n",
      "processing: Using machine learning approaches for multi-omics data analysis: A review\n",
      "processing: Machine learning-based identification of key lipid metabolism-related genes and associated ceRNA networks in NAFLD\n",
      "processing: A deep learning model for predicting long non-coding RNA and messenger RNA with model interpretation\n",
      "processing: AMPGLDA: Predicting LncRNA-Disease Associations Based on Adaptive Meta-Path Generation and Multi-Layer Perceptron\n",
      "processing: Performance of Existing Predictor of SubCellular Localization of Long Non-Coding RNA on the New Database\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/reza/miniconda3/envs/ml-lncRNA/lib/python3.12/site-packages/PyPDF2/_cmap.py:142: PdfReadWarning: Advanced encoding /UniGB-UTF16-H not implemented yet\n",
      "  warnings.warn(\n",
      "/home/reza/miniconda3/envs/ml-lncRNA/lib/python3.12/site-packages/PyPDF2/_cmap.py:142: PdfReadWarning: Advanced encoding /UniGB-UTF16-H not implemented yet\n",
      "  warnings.warn(\n",
      "/home/reza/miniconda3/envs/ml-lncRNA/lib/python3.12/site-packages/PyPDF2/_cmap.py:142: PdfReadWarning: Advanced encoding /UniGB-UTF16-H not implemented yet\n",
      "  warnings.warn(\n",
      "/home/reza/miniconda3/envs/ml-lncRNA/lib/python3.12/site-packages/PyPDF2/_cmap.py:142: PdfReadWarning: Advanced encoding /UniGB-UTF16-H not implemented yet\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "processing: … and Exploration of Immune Activation Pathways in T-cell Mediated Rejection through Integrated Bulk and Single-Cell RNA-Seq Analysis with Machine Learning\n",
      "processing: Machine Learning Models for Human Synapse Genomics\n",
      "processing: PREDICTING THE PRIMARY TISSUES OF CANCERS OF UNKNOWN PRIMARY USING MACHINE LEARNING\n",
      "processing: Application of deep learning in genomics\n",
      "processing: Machine learning-based diagnostic model of lymphatics-associated genes for new therapeutic target analysis in intervertebral disc degeneration\n",
      "processing: Common Features in lncRNA Annotation and Classification: A Survey. Noncoding RNA. 2021; 7: 77\n",
      "processing: SPIREX: Improving LLM-based relation extraction from RNA-focused scientific literature using graph machine learning\n",
      "processing: On the prediction of mRNA subcellular localization with machine learning\n",
      "processing: Machine learning methods for effectively discovering complex relationships in graph data\n",
      "processing: Mitochondrial Import of Malat1 Regulates Cardiac Mitochondrial Function in Type 2 Diabetes Mellitus through Interaction with MicroRNA-23b. Research Square\n",
      "processing: Impact of sequencing technologies on long non-coding RNA computational identification\n",
      "processing: Deciphering the methylation landscape in breast cancer: diagnostic and prognostic biosignatures through automated machine learning\n",
      "processing: Machine learning models for predicting lymph node and distant metastases in colorectal cancer\n",
      "processing: Development of New Bioinformatic Approaches for Human Genetic Studies\n"
     ]
    }
   ],
   "source": [
    "paper_terms = list()\n",
    "\n",
    "for paper_title, pdf_file_path in zip(paper_titles, paper_pdf_file_paths):\n",
    "    print(f'processing: {paper_title}')\n",
    "    result = search_in_file(pdf_file_path=pdf_file_path)\n",
    "    paper_terms.append({'paper_title': paper_title, 'result': result})    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d9ce381d-c0c0-4e5b-8f9d-0462eb5ee9eb",
   "metadata": {},
   "source": [
    "Flatten the search results and store as dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "e29e9533-5448-4158-9535-5669e55b7c64",
   "metadata": {},
   "outputs": [],
   "source": [
    "ml_lncRNA_search_result_df = pd.DataFrame([\n",
    "    (x['paper_title'], term, cnt)\n",
    "        for x in paper_terms\n",
    "            for term, cnt in x['result'].items()\n",
    "], columns = ['title', 'ml_term', 'cnt'])\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "49be44d0-38c2-4adc-bd24-dfeace56aa14",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>title</th>\n",
       "      <th>ml_term</th>\n",
       "      <th>cnt</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>LncMachine: a machine learning algorithm for l...</td>\n",
       "      <td>RANDOM_FORESTS</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>LncMachine: a machine learning algorithm for l...</td>\n",
       "      <td>SUPPORT_VECTOR_MACHINES</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>LncMachine: a machine learning algorithm for l...</td>\n",
       "      <td>LOGISTIC_REGRESSION</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>LncMachine: a machine learning algorithm for l...</td>\n",
       "      <td>DECISION_TREES</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LncMachine: a machine learning algorithm for l...</td>\n",
       "      <td>BOOSTING</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>496</th>\n",
       "      <td>Development of New Bioinformatic Approaches fo...</td>\n",
       "      <td>RANDOM_FORESTS</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>497</th>\n",
       "      <td>Development of New Bioinformatic Approaches fo...</td>\n",
       "      <td>SUPPORT_VECTOR_MACHINES</td>\n",
       "      <td>62</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>498</th>\n",
       "      <td>Development of New Bioinformatic Approaches fo...</td>\n",
       "      <td>PRINCIPAL_COMPONENT_ANALYSIS</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>499</th>\n",
       "      <td>Development of New Bioinformatic Approaches fo...</td>\n",
       "      <td>DECISION_TREES</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>500</th>\n",
       "      <td>Development of New Bioinformatic Approaches fo...</td>\n",
       "      <td>INDEPENDENT_COMPONENT_ANALYSIS</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>501 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 title  \\\n",
       "0    LncMachine: a machine learning algorithm for l...   \n",
       "1    LncMachine: a machine learning algorithm for l...   \n",
       "2    LncMachine: a machine learning algorithm for l...   \n",
       "3    LncMachine: a machine learning algorithm for l...   \n",
       "4    LncMachine: a machine learning algorithm for l...   \n",
       "..                                                 ...   \n",
       "496  Development of New Bioinformatic Approaches fo...   \n",
       "497  Development of New Bioinformatic Approaches fo...   \n",
       "498  Development of New Bioinformatic Approaches fo...   \n",
       "499  Development of New Bioinformatic Approaches fo...   \n",
       "500  Development of New Bioinformatic Approaches fo...   \n",
       "\n",
       "                            ml_term  cnt  \n",
       "0                    RANDOM_FORESTS    9  \n",
       "1           SUPPORT_VECTOR_MACHINES   11  \n",
       "2               LOGISTIC_REGRESSION    2  \n",
       "3                    DECISION_TREES    1  \n",
       "4                          BOOSTING    1  \n",
       "..                              ...  ...  \n",
       "496                  RANDOM_FORESTS   16  \n",
       "497         SUPPORT_VECTOR_MACHINES   62  \n",
       "498    PRINCIPAL_COMPONENT_ANALYSIS    1  \n",
       "499                  DECISION_TREES    1  \n",
       "500  INDEPENDENT_COMPONENT_ANALYSIS    1  \n",
       "\n",
       "[501 rows x 3 columns]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ml_lncRNA_search_result_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "b2606897-e9ff-46f8-b106-2a0ca63dd473",
   "metadata": {},
   "outputs": [],
   "source": [
    "ml_lncRNA_search_result_df.to_parquet(f'{DATA_FOLDER}ml_lncRNA_search_result_df.parquet')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "67b823e7-3877-407e-9c3d-3c0346104c6a",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
